### DISCUSSION TITLE
Keycloak OIDC Broker

### DISCUSSION THREAD

**Discussion by cyberjeff-u8t** [ubuntu/authd#689] [Feature Requests]
Is there any roadmap to add additional broker like Keycloak ?

**Comment by mmendozaf** [comment]
This is a good request as KeyCloack is widely

**Comment by adombeck** [comment]
It's already possible to use Keycloak via the [authd-google](https://snapcraft.io/authd-google) broker (which really is just a generic OpenID Connect broker).

You need to set up a client in Keycloak, choose a client ID and enable `OAuth 2.0 Device Authorization Grant`. Then edit `/var/snap/authd-google/current/broker.conf` and set:
```
[oidc]
issuer = <your-keycloak-domain>/realms/<your-realm>
client_id = <your-client-ID>
```

Then restart the broker:
```
sudo snap restart authd-google
```

You should then be able to log in with any user of the configured Keycloak realm.

**Reply by mueckinger** [reply]
I tried this but it didn't work. With max. verbosity of authd-google logs, there is nothing to be gained from it:
```
Nov 14 14:08:34 nuc authd-google[49713]: context canceled
Nov 14 14:08:34 nuc authd-google[49713]: IsAuthenticated error: context canceled
Nov 14 14:08:34 nuc authd[47195]: can't check authentication: invalid access authentication key:
```

Could you provide more information on your setup? Also on the Keycloak settings?

**Reply by adombeck** [reply]
Those logs are incomplete. Please file a GitHub issue via https://github.com/ubuntu/authd/issues/new/choose with the logs so I can debug the issue.

**Reply by mueckinger** [reply]
Sure. Could you help me on how to increase verbosity? I am already at `/usr/bin/snap run authd-google -vvv`

**Reply by adombeck** [reply]
`-vvv` is the max verbosity for the broker. You can also increase the verbosity for authd though. See https://documentation.ubuntu.com/authd/stable-docs/howto/logging/#configure-logging-verbosity

**Reply by mueckinger** [reply]
authd logs with `-vvv`:
```
Nov 14 14:41:08 nuc (authd)[58238]: Creating user manager with config: {UIDMin:1000000000 UIDMax:1999999999 GIDMin:1000000000 GIDMax:1999999999}
Nov 14 14:41:08 nuc (authd)[58238]: Building new gRPC user service
Nov 14 14:41:08 nuc (authd)[58238]: Building new gRPC PAM service
Nov 14 14:41:08 nuc (authd)[58238]: Building new daemon
Nov 14 14:41:08 nuc (authd)[58238]: Use socket activation
Nov 14 14:41:08 nuc (authd)[58238]: Registering gRPC services
Nov 14 14:41:08 nuc (authd)[58238]: Starting to serve requests on /run/authd.sock
Nov 14 14:41:08 nuc (authd)[58238]: Ready state sent to systemd
Nov 14 14:41:08 nuc (authd)[58238]: Serving gRPC requests on /run/authd.sock
Nov 14 14:41:08 nuc systemd[1]: Started authd.service - Authd daemon service.
Nov 14 14:42:40 nuc authd[58238]: Pre-checking user "<redacted>"
Nov 14 14:42:40 nuc authd[58238]: Added temporary record for user "<redacted>" with UID 1012281109
Nov 14 14:42:40 nuc authd[58238]: GetUserByName: no result matching  in 
Nov 14 14:42:40 nuc authd[58238]: rpc error: code = NotFound desc = 
Nov 14 14:42:40 nuc authd[58238]: User "<redacted>" is unknown
Nov 14 14:42:40 nuc authd[58238]: Pre-checking user "<redacted>"
Nov 14 14:42:44 nuc authd[58238]: 3820511142-e4b86797-4e2a-4de6-a40a-b1357221adc8: New auth session for "<redacted>"
Nov 14 14:44:40 nuc authd[58238]: can't check authentication: invalid access authentication key: 
```

**Reply by adombeck** [reply]
I need more information, for example the authd and broker versions and their config files. please file an issue and use the command in the issue template to retrieve all the required information and logs.

**Reply by adombeck** [reply]
You can also try with the edge versions of authd and the broker: https://documentation.ubuntu.com/authd/stable-docs/howto/changing-versions/#ref-changing-versions

**Reply by mueckinger** [reply]
Thank you. But I initially asked for documentation of your Keycloak setup. I cannot rule out that the issue is a wrong setup on Keycloak's side since it has a myriad of possible settings. I have assumed from your comment that you have tested authd with Keycloak.

**Reply by mueckinger** [reply]
Is there a possibility to view and debug the backchannel communication between authd and the IdP?

**Reply by adombeck** [reply]
> But I initially asked for documentation of your Keycloak setup. 

I started Keycloak via

```
sudo docker run -p 127.0.0.1:8080:8080 -e KC_BOOTSTRAP_ADMIN_USERNAME=admin -e KC_BOOTSTRAP_ADMIN_PASSWORD=admin -e KC_HTTP_ENABLED=true -e KC_HOSTNAME_STRICT=false -e KC_HOSTNAME_STRICT_HTTPS=false quay.io/keycloak/keycloak:latest start-dev
```

I don't think I changed anything else than what I wrote above:
> You need to set up a client in Keycloak, choose a client ID and enable `OAuth 2.0 Device Authorization Grant`. Then edit `/var/snap/authd-google/current/broker.conf` and set:
> 
> ```
> [oidc]
> issuer = <your-keycloak-domain>/realms/<your-realm>
> client_id = <your-client-ID>
> ```

> I cannot rule out that the issue is a wrong setup on Keycloak's side since it has a myriad of possible settings.

That's fine, I can still take a look. But with the limited logs you pasted here I can't help. If you create the GitHub issue and it turns out to not be an issue in authd we can just close the issue, that's no problem.
